transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,579 @@
+
+use futures::{
+    channel::{mpsc, oneshot},
+    lock::Mutex,
+    prelude::*,
+    stream::Stream,
+};
+use libp2p_core::multiaddr::Multiaddr;
+use quinn_proto::{ClientConfig as QuinnClientConfig, ServerConfig as QuinnServerConfig};
+use std::{
+    collections::{HashMap, VecDeque},
+    fmt, io,
+    pin::Pin,
+    sync::{Arc, Weak},
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+/// Represents the configuration for the [`Endpoint`].
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The client configuration to pass to `quinn_proto`.
+    client_config: quinn_proto::ClientConfig,
+    /// The server configuration to pass to `quinn_proto`.
+    server_config: Arc<quinn_proto::ServerConfig>,
+    /// The endpoint configuration to pass to `quinn_proto`.
+    endpoint_config: Arc<quinn_proto::EndpointConfig>,
+    /// The [`Multiaddr`] to use to spawn the UDP socket.
+    multiaddr: Multiaddr,
+}
+
+impl Config {
+    /// Creates a new configuration object with default values.
+    pub fn new(
+        keypair: &libp2p_core::identity::Keypair,
+        multiaddr: Multiaddr,
+    ) -> Result<Self, tls::ConfigError> {
+        let mut transport = quinn_proto::TransportConfig::default();
+        transport.max_concurrent_uni_streams(0u32.into()); // Can only panic if value is out of range.
+        transport.datagram_receive_buffer_size(None);
+        transport.keep_alive_interval(Some(Duration::from_millis(10)));
+        let transport = Arc::new(transport);
+
+        let client_tls_config = tls::make_client_config(keypair).unwrap();
+        let server_tls_config = tls::make_server_config(keypair).unwrap();
+
+        let mut server_config = QuinnServerConfig::with_crypto(Arc::new(server_tls_config));
+        server_config.transport = Arc::clone(&transport);
+
+        let mut client_config = QuinnClientConfig::new(Arc::new(client_tls_config));
+        client_config.transport = transport;
+        Ok(Self {
+            client_config,
+            server_config: Arc::new(server_config),
+            endpoint_config: Default::default(),
+            multiaddr,
+        })
+    }
+}
+
+/// Object containing all the QUIC resources shared between all connections.
+// TODO: expand docs
+// TODO: Debug trait
+// TODO: remove useless fields
+pub struct Endpoint {
+    /// Channel to the background of the endpoint.
+    /// See [`Endpoint::new_connections`] (just below) for a commentary about the mutex.
+    to_endpoint: Mutex<mpsc::Sender<ToEndpoint>>,
+
+    /// Channel where new connections are being sent.
+    /// This is protected by a futures-friendly `Mutex`, meaning that receiving a connection is
+    /// done in two steps: locking this mutex, and grabbing the next element on the `Receiver`.
+    /// The only consequence of this `Mutex` is that multiple simultaneous calls to
+    /// [`Endpoint::next_incoming`] are serialized.
+    new_connections: Mutex<mpsc::Receiver<Connection>>,
```

Comment:
Having the receiver in a Mutex theoretically introduces a race condition when polling it. Now I know that it is only used in a single place, that is `Endpoint::poll_incoming`, which is only called in the `QuicTransport`. Nevertheless I think we should ensure on a type level that this will never be called from two different places. Hence I would propose to hold the `mpsc::Receiver<Connection>` directly in the `QuicTransport` and poll it there directly.